source/common/upstream/original_dst_cluster.cc

```diff
@@ -40,6 +41,23 @@ OriginalDstCluster::LoadBalancer::LoadBalancer(PrioritySet& priority_set, Cluste
 
 HostConstSharedPtr OriginalDstCluster::LoadBalancer::chooseHost(LoadBalancerContext* context) {
   if (context) {
+    // Check if request has a override host header set, if directly return the host using it.
+    Envoy::Router::Filter* filter = dynamic_cast<Envoy::Router::Filter*>(context);
```

Comment:
@mattklein123  I am actually adding a method downStreamHeaders in the LoadBalancerContext so that it can be easily retrieved here with out down_cast and also is much cleaner and easier for tests as well.